transport/http/handler.go

```diff
@@ -62,13 +63,48 @@ func (h handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	service := popHeader(req.Header, ServiceHeader)
 	procedure := popHeader(req.Header, ProcedureHeader)
 	bothResponseError := popHeader(req.Header, AcceptsBothResponseErrorHeader) == AcceptTrue
-	// add response header to echo accepted rpc-service
+
+	// Extract tracing context and remove tracing headers before they reach handlers
+	carrier := opentracing.HTTPHeadersCarrier(req.Header)
+	parentSpanCtx, _ := h.tracer.Extract(opentracing.HTTPHeaders, carrier)
+
+	// Remove tracing headers so application handlers don't see them
+	for k := range req.Header {
+		kLower := strings.ToLower(k)
+		if kLower == strings.ToLower(UberTraceIDHeader) || strings.HasPrefix(kLower, strings.ToLower(UberCtxHeader)) {
+			req.Header.Del(k)
+		}
+	}
+
+	// Start span
+	start := time.Now()
+	tags := opentracing.Tags{
+		"rpc.service":   service,
+		"rpc.transport": "http",
+	}
+	for k, v := range yarpc.OpentracingTags {
+		tags[k] = v
+	}
+	span := h.tracer.StartSpan(
+		procedure,
+		opentracing.StartTime(start),
+		ext.RPCServerOption(parentSpanCtx),
+		tags,
+	)
+	ext.PeerService.Set(span, service)
+	ctx := opentracing.ContextWithSpan(req.Context(), span)
```

Comment:
We should not create second context, request's context should be replaced with req.WithContext.